### PR TITLE
added docker-build-push github action

### PR DESCRIPTION
**Pull request by RonaldCrb**
This PR configures a github action that builds and push the lnd image to docker-hub based on any push to a branch with a name like => `0.*` ; 

for this action to work, it requires the following secrets to be configured in the repo.  

1. `DOCKERHUB_USERNAME`
2. `DOCKERHUB_TOKEN`

### CODE REVIEW

File: .github/workflows/docker-build-push.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+      
+      - name: Build and push
+        id: docker_build
+        uses: docker/build-push-action@v2
+        with:
+          push: true
+          tags: user/app:latest
```
Comment (mrfelton) [inline .github/workflows/docker-build-push.yaml:30]: Presumably we need to set this to the correct image name here?

**Review by mrfelton — CHANGES_REQUESTED** [review]

File: .github/workflows/docker-build-push.yaml (hunk 1 above)
Comment (RonaldCrb) [reply .github/workflows/docker-build-push.yaml:30]: ahhh youre right. i forgot to change that

**Comment by mrfelton** [comment]
Can we integrate this with tags too @RonaldCrb  such that pushing a github tag triggers a build and creation of image with said tag?

**Comment by RonaldCrb** [comment]
thats a great idea, i didnt thought about the tags in this repo because i saw there were none. i will look into it, thanks for the quick feedback!

**Comment by mrfelton** [comment]
> i didnt thought about the tags in this repo because i saw there were none

Yeah, that's because this repo has not had the tlc that it needs to date. Today is a good day to change that :)

**Comment by RonaldCrb** [comment]
> Looks good!
> 
> Only thing we need before merge is a test that makes sure the build is in good state before publish - something as simple as `lncli --version` (expecting a match of whatever version we're building) should suffice.

im investigating how to achieve this since the action im proposing builds and pushes the image in a single motion

**Comment by mrfelton** [comment]
Another thing that would be very useful would be to also build and push a debug version of the image. Meaning, a version that bundles the debug binaries rather than the production ones.

You can see how to do this in the branch https://github.com/LN-Zap/docker-lnd/tree/0.10.x-beta-debug where I did it once before. The key think is to copy the debug binaries into the container rather than the normal ones, as per https://github.com/LN-Zap/docker-lnd/blob/0.10.x-beta-debug/Dockerfile#L22-L23

This is sometimes needed in order to resolve issues like the one described in https://github.com/LN-Zap/zap-middleware/issues/2460 (one way to resolve these are to run a debug version of lnd - which adds the `abandonchannel` rpc - and then to call that).

Not sure if you want to handle that here or in a follow up PR though.

**Comment by RonaldCrb** [comment]
this is ready for review, 

added:

1. build the image with each release
2. build both the normal and debug images and tag them with the release tag
3. smoke test on both images using `lncli --version` and comparing it to the release tag
4. push the images to dockerhub

there is an outstanding point concerning the update of the readme file on the dockerhub page, so i did commented that section of the pipeline definition until we can provide the `zap-ci` user with privileges to update the repo metadata and the readme page of the images

File: .github/scripts/smoke-test.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+set -e
+
+echo "release tag   => $1"
+echo "lncli version => $2"
+
+if [ $1 == $2 ]
+then
+  exit 0
+else
+  exit 42
```
Comment (szollo) [inline .github/scripts/smoke-test.sh:10]: Is there a reasoning why this is `exit 42` instead of a normal `exit 1`?

Comment (RonaldCrb) [reply .github/scripts/smoke-test.sh]: there is no objective reasoning behind that. in any case i have changed it and also added meaningful messages for success and error.

**Review by szollo — APPROVED** [review]
LGTM!

So I guess our new workflow to build lnd images in this repo from now on will be to tag each new release using GitHub Releases.

**Comment by RonaldCrb** [comment]
> LGTM!
> 
> So I guess our new workflow to build lnd images in this repo from now on will be to tag each new release using GitHub Releases.

correct, each release tag becomes the docker image tag, so in an ideal situation, one will create a branch, compile the image locally to make sure the desired result is achieved, then create a release tag. once the release is published, the action will trigger. and the image is actually pushed to dockerhub.

**Pull request merged**